### PR TITLE
[SVLS-4699] Detect exceptions raised outside of the handler

### PR DESCRIPTION
**Pull request by TalUsvyatsky**
### What does this PR do?

This PR adds telemetry for exceptions that occur outside of the handler function. 

When an exception occurs outside of the handler function:
1. `aws.lambda.enhanced.errors` is emitted
2. A trace with an error span is created

### Motivation

- [SVLS-4699](https://datadoghq.atlassian.net/browse/SVLS-4699)
- Users should have visibility into this category of errors
- We did the same thing in Python in https://github.com/DataDog/datadog-lambda-python/pull/475

### Testing Guidelines

1. Created a handler file like this for testing "cjs" handler:
```js
  throw new TypeError("im a type error")
  exports.handler = function async (event) {
    const response = {
      statusCode: 200,
      body: JSON.stringify('Hello from Lambda!'),
    };
    return response;
  };
```
and one like this for testing "mjs" handler:
```js
  throw new TypeError("im a type error")
  export const handler = async (event) => {
    const response = {
      statusCode: 200,
      body: JSON.stringify('Hello from Lambda!'),
    };
    return response;
  };
```

2. Deployed my function using each handler and invoked it
3. Saw traces with an error span from the call ([example trace](https://ddserverless.datadoghq.com/apm/trace/811334263615813462?colorBy=service&env=dev&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=811334263615813462&spanViewType=errors&timeHint=1715095582277))
4. Saw `aws.lambda.enhanced.errors` emitted ([graph](https://ddserverless.datadoghq.com/s/61ba5v0dnxcpd8dm/nwd-nag-6uj))
5. Traces appear in serverless detail page invocation table ([link to search for function](https://ddserverless.datadoghq.com/functions?search=region%3Asa-east-1&cloud=aws&entity_view=lambda_functions&fromUser=false&panel_end=1715095740000&panel_paused=false&panel_start=1715094840000&panel_tab=invocations&shouldShowLegend=true&sort=time&sort_lambda=-name&sp=%5B%7B%22p%22%3A%7B%22entityId%22%3A%22aws-lambda-functions%2Btu-node-dev-function1%2Bsa-east-1%2B425362996713%22%7D%2C%22i%22%3A%22lambda-panel%22%7D%5D&spanViewType=errors&text_search=tu-&view=spans&start=1715081340000&end=1715095740000&paused=false))
6. Issues appear in error tracking and are informative

### Types of Changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [x] This PR passes the integration tests (ask a Datadog member to run the tests)

[SVLS-4699]: https://datadoghq.atlassian.net/browse/SVLS-4699?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/handler.cjs [hunk 1]
```diff
@@ -32,4 +33,9 @@ if (extractorEnv) {
   }
 }
 
-exports.handler = datadog(loadSync(taskRootEnv, handlerEnv), { traceExtractor });
+try {
+  exports.handler = datadog(loadSync(taskRootEnv, handlerEnv), { traceExtractor });
+} catch (error) {
+  emitTelemetryOnErrorOutsideHandler(error, handlerEnv, Date.now());
```
Comment (joeyzhao2018) [inline src/handler.cjs:39]: I might be wrong but is this `fire and forget` way of invoking the async method? I'm just wondering if we need to add `.then()` or `.error()` to make sure it ends?

File: src/handler.mjs [hunk 2]
```diff
@@ -26,4 +33,12 @@ if (extractorEnv) {
   }
 }
 
-export const handler = datadog(await load(taskRootEnv, handlerEnv), { traceExtractor });
+let wrapped_handler;
```
Comment (purple4reina) [inline src/handler.mjs:36]: nit: use camel case `wrappedHandler`

File: src/metrics/enhanced-metrics.ts [hunk 3]
```diff
@@ -49,12 +49,17 @@ export function getRuntimeTag(): string | null {
...
-  if (context.invokedFunctionArn) {
-    arnTags = parseTagsFromARN(context.invokedFunctionArn, context.functionVersion);
+export function getEnhancedMetricTags(context?: Context): string[] {
+  let tags: string[] = [];
+  if (context) {
+    let arnTags = [`functionname:${context.functionName}`];
+    if (context.invokedFunctionArn) {
+      arnTags = parseTagsFromARN(context.invokedFunctionArn, context.functionVersion);
+    }
+    tags = [...arnTags, `memorysize:${context.memoryLimitInMB}`];
```
Comment (purple4reina) [inline src/metrics/enhanced-metrics.ts:59]: Does js have an `append`-like function? so we don't have to do the extra allocation.

File: src/handler.cjs (hunk 1 above)
Comment (TalUsvyatsky) [reply src/handler.cjs]: thanks! yes I think you're right, added

**Review by joeyzhao2018 — APPROVED** [review]

File: src/index.ts [hunk 4]
```diff
@@ -416,3 +418,37 @@ function getRuntimeTag(): string {
...
+    span.finish();
+  }
+
+  const config = getConfig();
+  const metricsListener = new MetricsListener(new KMSService(), config);
+  await metricsListener.onStartInvocation(undefined);
+  if (config.enhancedMetrics) {
+    incrementErrorsMetric(metricsListener);
+  }
+  await metricsListener.onCompleteInvocation();
```
Comment (duncanista) [inline src/index.ts:453]: Why don't we just lazy load everything inside of the condition?

File: src/metrics/enhanced-metrics.ts [hunk 5]
```diff
@@ -49,12 +49,17 @@ export function getRuntimeTag(): string | null {
...
 
-export function getEnhancedMetricTags(context: Context): string[] {
-  let arnTags = [`functionname:${context.functionName}`];
-  if (context.invokedFunctionArn) {
-    arnTags = parseTagsFromARN(context.invokedFunctionArn, context.functionVersion);
+export function getEnhancedMetricTags(context?: Context): string[] {
+  const tags: string[] = [];
+  if (context) {
+    let arnTags = [`functionname:${context.functionName}`];
+    if (context.invokedFunctionArn) {
```
Comment (duncanista) [inline src/metrics/enhanced-metrics.ts:56]: I think it's guaranteed that the `context.invokeFunctionArn` [is going to be present](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html).

Do we really need the condition?

**Review by duncanista — COMMENTED** [review]
Left some comments.

File: src/metrics/enhanced-metrics.ts (hunk 5 above)
Comment (duncanista) [reply src/metrics/enhanced-metrics.ts:56]: Thanks for clarification

**Review by duncanista — APPROVED** [review]

**Pull request merged**